File: backend/session/compaction_utils.go
package session

import (
	"encoding/json"
	"fmt"
	"sort"
	"strings"

	"github.com/enough/enough/backend/config"
	"github.com/enough/enough/backend/opencode"
)

const (
	CompactionSummaryPrefix = "The conversation history before this point was compacted into the following summary:\n\n<summary>\n"
	CompactionSummarySuffix = "\n</summary>"
	BranchSummaryPrefix     = "The following is a summary of a branch that this conversation came back from:\n\n<summary>\n"
	BranchSummarySuffix     = "\n</summary>"
)

type FileOperations struct {
	Read    map[string]bool
	Written map[string]bool
	Edited  map[string]bool
}

func NewFileOps() FileOperations {
	return FileOperations{
		Read:    make(map[string]bool),
		Written: make(map[string]bool),
		Edited:  make(map[string]bool),
	}
}

type CompactionDetails struct {
	ReadFiles     []string `json:"readFiles"`
	ModifiedFiles []string `json:"modifiedFiles"`
}

func ExtractFileOpsFromMessage(msg opencode.Message, fileOps FileOperations) {
	if msg.Role != "assistant" {
		return
	}
	for _, tc := range msg.ToolCalls {
		var args map[string]any
		if err := json.Unmarshal([]byte(tc.Function.Arguments), &args); err != nil {
			continue
		}
		pathVal, ok := args["path"]
		if !ok {
			continue
		}
		path, ok := pathVal.(string)
		if !ok || path == "" {
			continue
		}
		switch tc.Function.Name {
		case "read_file":
			fileOps.Read[path] = true
		case "write_file":
			fileOps.Written[path] = true
		case "edit_file":
			fileOps.Edited[path] = true
		}
	}
}

func ComputeFileLists(fileOps FileOperations) ([]string, []string) {
	modified := make(map[string]bool)
	for f := range fileOps.Edited {
		modified[f] = true
	}
	for f := range fileOps.Written {
		modified[f] = true
	}

	var readOnly []string
	for f := range fileOps.Read {
		if !modified[f] {
			readOnly = append(readOnly, f)
		}
	}
	sort.Strings(readOnly)

	var modifiedFiles []string
	for f := range modified {
		modifiedFiles = append(modifiedFiles, f)
	}
	sort.Strings(modifiedFiles)

	return readOnly, modifiedFiles
}

func FormatFileOperations(readFiles []string, modifiedFiles []string) string {
	var sections []string
	if len(readFiles) > 0 {
		sections = append(sections, fmt.Sprintf("<read-files>\n%s\n</read-files>", strings.Join(readFiles, "\n")))
	}
	if len(modifiedFiles) > 0 {
		sections = append(sections, fmt.Sprintf("<modified-files>\n%s\n</modified-files>", strings.Join(modifiedFiles, "\n")))
	}
	if len(sections) == 0 {
		return ""
	}
	return "\n\n" + strings.Join(sections, "\n\n")
}

func truncateForSummary(text string, maxChars int) string {
	if len(text) <= maxChars {
		return text
	}
	truncatedChars := len(text) - maxChars
	return fmt.Sprintf("%s\n\n[... %d more characters truncated]", text[:maxChars], truncatedChars)
}

func SerializeConversation(messages []opencode.Message) string {
	var parts []string
	for _, msg := range messages {
		if msg.Role == "user" {
			content := opencode.ContentString(msg)
			if content != "" {
				parts = append(parts, fmt.Sprintf("[User]: %s", content))
			}
		} else if msg.Role == "assistant" {
			var textParts []string
			var thinkingParts []string
			var toolCalls []string

			if msg.ReasoningContent != nil && *msg.ReasoningContent != "" {
				thinkingParts = append(thinkingParts, *msg.ReasoningContent)
			}
			content := opencode.ContentString(msg)
			if content != "" {
				textParts = append(textParts, content)
			}
			for _, tc := range msg.ToolCalls {
				var args map[string]any
				var argsStr string
				if json.Unmarshal([]byte(tc.Function.Arguments), &args) == nil {
					var kv []string
					// Sort keys for deterministic serialization in tests
					var keys []string
					for k := range args {
						keys = append(keys, k)
					}
					sort.Strings(keys)
					for _, k := range keys {
						v := args[k]
						vBytes, _ := json.Marshal(v)
						kv = append(kv, fmt.Sprintf("%s=%s", k, string(vBytes)))
					}
					argsStr = strings.Join(kv, ", ")
				} else {
					argsStr = tc.Function.Arguments
				}
				toolCalls = append(toolCalls, fmt.Sprintf("%s(%s)", tc.Function.Name, argsStr))
			}

			if len(thinkingParts) > 0 {
				parts = append(parts, fmt.Sprintf("[Assistant thinking]: %s", strings.Join(thinkingParts, "\n")))
			}
			if len(textParts) > 0 {
				parts = append(parts, fmt.Sprintf("[Assistant]: %s", strings.Join(textParts, "\n")))
			}
			if len(toolCalls) > 0 {
				parts = append(parts, fmt.Sprintf("[Assistant tool calls]: %s", strings.Join(toolCalls, "; ")))
			}
		} else if msg.Role == "tool" {
			content := opencode.ContentString(msg)
			if content != "" {
				parts = append(parts, fmt.Sprintf("[Tool result]: %s", truncateForSummary(content, 2000)))
			}
		}
	}
	return strings.Join(parts, "\n\n")
}

func ConvertToLlm(messages []opencode.Message) []opencode.Message {
	var out []opencode.Message
	for _, m := range messages {
		switch m.Role {
		case "compactionSummary":
			content := CompactionSummaryPrefix + opencode.ContentString(m) + CompactionSummarySuffix
			out = append(out, opencode.Message{
				Role:    "user",
				Content: opencode.StringContent(content),
			})
		case "branchSummary":
			content := BranchSummaryPrefix + opencode.ContentString(m) + BranchSummarySuffix
			out = append(out, opencode.Message{
				Role:    "user",
				Content: opencode.StringContent(content),
			})
		default:
			out = append(out, m)
		}
	}
	return out
}

func CalculateContextTokens(usage opencode.Usage) int {
	if usage.TotalTokens > 0 {
		return usage.TotalTokens
	}
	return usage.Input + usage.Output + usage.CacheRead + usage.CacheWrite
}

func GetLastAssistantUsage(entries []FileEntry) *opencode.Usage {
	for i := len(entries) - 1; i >= 0; i-- {
		if entries[i].Type == TypeMessage && entries[i].Message != nil && entries[i].Message.Role == "assistant" {
			if entries[i].Message.Usage != nil {
				return entries[i].Message.Usage
			}
		}
	}
	return nil
}

type ContextUsageEstimate struct {
	Tokens         int
	UsageTokens    int
	TrailingTokens int
	LastUsageIndex int
}

func EstimateMessageTokens(msg opencode.Message) int {
	chars := 0
	if msg.Role == "user" {
		if len(msg.Content) > 0 {
			chars = len(opencode.ContentString(msg))
		}
		return (chars + 3) / 4
	}
	if msg.Role == "assistant" {
		if len(msg.Content) > 0 {
			chars += len(opencode.ContentString(msg))
		}
		if msg.ReasoningContent != nil {
			chars += len(*msg.ReasoningContent)
		}
		for _, tc := range msg.ToolCalls {
			chars += len(tc.Function.Name) + len(tc.Function.Arguments)
		}
		return (chars + 3) / 4
	}
	if msg.Role == "tool" || msg.Role == "toolResult" {
		if len(msg.Content) > 0 {
			chars = len(opencode.ContentString(msg))
		}
		return (chars + 3) / 4
	}
	if msg.Role == "compactionSummary" || msg.Role == "branchSummary" {
		if len(msg.Content) > 0 {
			chars = len(opencode.ContentString(msg))
		}
		return (chars + 3) / 4
	}
	return 0
}

func EstimateTokens(entry FileEntry) int {
	if entry.Type == TypeCompaction || entry.Type == TypeBranchSummary {
		return (len(entry.Summary) + 3) / 4
	}
	if entry.Type != TypeMessage || entry.Message == nil {
		return 0
	}
	return EstimateMessageTokens(*entry.Message)
}

func EstimateContextTokens(messages []opencode.Message) ContextUsageEstimate {
	lastUsageIdx := -1
	var lastUsage *opencode.Usage
	for i := len(messages) - 1; i >= 0; i-- {
		if messages[i].Role == "assistant" && messages[i].Usage != nil {
			lastUsageIdx = i
			lastUsage = messages[i].Usage
			break
		}
	}

	if lastUsageIdx == -1 {
		estimated := 0
		for _, msg := range messages {
			estimated += EstimateMessageTokens(msg)
		}
		return ContextUsageEstimate{
			Tokens:         estimated,
			UsageTokens:    0,
			TrailingTokens: estimated,
			LastUsageIndex: -1,
		}
	}

	usageTokens := CalculateContextTokens(*lastUsage)
	trailingTokens := 0
	for i := lastUsageIdx + 1; i < len(messages); i++ {
		trailingTokens += EstimateMessageTokens(messages[i])
	}

	return ContextUsageEstimate{
		Tokens:         usageTokens + trailingTokens,
		UsageTokens:    usageTokens,
		TrailingTokens: trailingTokens,
		LastUsageIndex: lastUsageIdx,
	}
}

func ShouldCompact(contextTokens int, contextWindow int, settings config.CompactionSettings) bool {
	if !settings.Enabled {
		return false
	}
	return contextTokens > contextWindow-settings.ReserveTokens
}

func findValidCutPoints(entries []FileEntry, startIndex int, endIndex int) []int {
	var cutPoints []int
	for i := startIndex; i < endIndex; i++ {
		entry := entries[i]
		if entry.Type == TypeMessage && entry.Message != nil {
			role := entry.Message.Role
			if role == "user" || role == "assistant" || role == "bashExecution" || role == "custom" || role == "branchSummary" || role == "compactionSummary" {
				cutPoints = append(cutPoints, i)
			}
		} else if entry.Type == TypeBranchSummary || entry.Type == TypeCustomMessage {
			cutPoints = append(cutPoints, i)
		}
	}
	return cutPoints
}

func findTurnStartIndex(entries []FileEntry, entryIndex int, startIndex int) int {
	for i := entryIndex; i >= startIndex; i-- {
		entry := entries[i]
		if entry.Type == TypeBranchSummary || entry.Type == TypeCustomMessage {
			return i
		}
		if entry.Type == TypeMessage && entry.Message != nil {
			role := entry.Message.Role
			if role == "user" || role == "bashExecution" {
				return i
			}
		}
	}
	return -1
}

type CutPointResult struct {
	FirstKeptEntryIndex int
	TurnStartIndex      int
	IsSplitTurn         bool
}

func FindCutPoint(entries []FileEntry, startIndex int, endIndex int, keepRecentTokens int) CutPointResult {
	cutPoints := findValidCutPoints(entries, startIndex, endIndex)
	if len(cutPoints) == 0 {
		return CutPointResult{FirstKeptEntryIndex: startIndex, TurnStartIndex: -1, IsSplitTurn: false}
	}

	accumulatedTokens := 0
	cutIndex := cutPoints[0]

	for i := endIndex - 1; i >= startIndex; i-- {
		entry := entries[i]
		if entry.Type != TypeMessage {
			continue
		}
		accumulatedTokens += EstimateTokens(entry)
		if accumulatedTokens >= keepRecentTokens {
			for _, cp := range cutPoints {
				if cp >= i {
					cutIndex = cp
					break
				}
			}
			break
		}
	}

	for cutIndex > startIndex {
		prevEntry := entries[cutIndex-1]
		if prevEntry.Type == TypeCompaction || prevEntry.Type == TypeMessage {
			break
		}
		cutIndex--
	}

	cutEntry := entries[cutIndex]
	isUserMessage := cutEntry.Type == TypeMessage && cutEntry.Message != nil && cutEntry.Message.Role == "user"
	turnStartIndex := -1
	if !isUserMessage {
		turnStartIndex = findTurnStartIndex(entries, cutIndex, startIndex)
	}

	return CutPointResult{
		FirstKeptEntryIndex: cutIndex,
		TurnStartIndex:      turnStartIndex,
		IsSplitTurn:         !isUserMessage && turnStartIndex != -1,
	}
}

// GetLatestCompactionEntry walks the branch backwards and returns the latest compaction entry, if any.
func GetLatestCompactionEntry(branch []FileEntry) *FileEntry {
	for i := len(branch) - 1; i >= 0; i-- {
		if branch[i].Type == TypeCompaction {
			return &branch[i]
		}
	}
	return nil
}

File: backend/session/compaction_utils.ts
// PORT: mirrors backend/session/compaction_utils.go

import type { message, usage } from "../opencode/types";
import { string_content, content_string, content_blocks } from "../opencode/types";
import type { compaction_settings } from "../config/config";
import {
  type_message,
  type_compaction,
  type_branch_summary,
  type_custom_message,
  type file_entry,
} from "./types";

export const CompactionSummaryPrefix =
  "The conversation history before this point was compacted into the following summary:\n\n<summary>\n";
export const CompactionSummarySuffix = "\n</summary>";
export const BranchSummaryPrefix =
  "The following is a summary of a branch that this conversation came back from:\n\n<summary>\n";
export const BranchSummarySuffix = "\n</summary>";

export type file_operations = {
  read: Record<string, boolean>;
  written: Record<string, boolean>;
  edited: Record<string, boolean>;
};

export const new_file_ops = (): file_operations => ({
  read: {},
  written: {},
  edited: {},
});

export type compaction_details = {
  readFiles: string[];
  modifiedFiles: string[];
};

export const extract_file_ops_from_message = (
  msg: message,
  fileOps: file_operations,
): void => {
  if (msg.role !== "assistant") {
    return;
  }
  if (!msg.tool_calls) {
    return;
  }
  for (const tc of msg.tool_calls) {
    let args: any;
    try {
      args = JSON.parse(tc.function.arguments);
    } catch {
      continue;
    }
    if (!args || typeof args !== "object") {
      continue;
    }
    const pathVal = args["path"];
    if (typeof pathVal !== "string" || pathVal === "") {
      continue;
    }
    switch (tc.function.name) {
      case "read_file":
        fileOps.read[pathVal] = true;
        break;
      case "write_file":
        fileOps.written[pathVal] = true;
        break;
      case "edit_file":
        fileOps.edited[pathVal] = true;
        break;
    }
  }
};

export const extract_file_operations = (
  messages: message[],
  entries: file_entry[],
  prevCompactionIndex: number,
): file_operations => {
  const fileOps = new_file_ops();
  if (prevCompactionIndex >= 0) {
    const prevComp = entries[prevCompactionIndex];
    if (!prevComp.fromHook && prevComp.details) {
      const details = prevComp.details;
      if (details && typeof details === "object") {
        if (Array.isArray(details.readFiles)) {
          for (const f of details.readFiles) {
            if (typeof f === "string") {
              fileOps.read[f] = true;
            }
          }
        }
        if (Array.isArray(details.modifiedFiles)) {
          for (const f of details.modifiedFiles) {
            if (typeof f === "string") {
              fileOps.edited[f] = true;
            }
          }
        }
      }
    }
  }
  for (const msg of messages) {
    extract_file_ops_from_message(msg, fileOps);
  }
  return fileOps;
};

export const compute_file_lists = (fileOps: file_operations): [string[], string[]] => {
  const modified: Record<string, boolean> = {};
  for (const f of Object.keys(fileOps.edited)) {
    modified[f] = true;
  }
  for (const f of Object.keys(fileOps.written)) {
    modified[f] = true;
  }

  const readOnly: string[] = [];
  for (const f of Object.keys(fileOps.read)) {
    if (!modified[f]) {
      readOnly.push(f);
    }
  }
  readOnly.sort();

  const modifiedFiles = Object.keys(modified);
  modifiedFiles.sort();

  return [readOnly, modifiedFiles];
};

export const format_file_operations = (
  readFiles: string[],
  modifiedFiles: string[],
): string => {
  const sections: string[] = [];
  if (readFiles.length > 0) {
    sections.push(`<read-files>\n${readFiles.join("\n")}\n</read-files>`);
  }
  if (modifiedFiles.length > 0) {
    sections.push(`<modified-files>\n${modifiedFiles.join("\n")}\n</modified-files>`);
  }
  if (sections.length === 0) {
    return "";
  }
  return "\n\n" + sections.join("\n\n");
};

const safe_json_stringify = (value: unknown): string => {
  try {
    return JSON.stringify(value) ?? "undefined";
  } catch {
    return "[unserializable]";
  }
};

const truncate_for_summary = (text: string, maxChars: number): string => {
  if (text.length <= maxChars) {
    return text;
  }
  const truncatedChars = text.length - maxChars;
  return `${text.slice(0, maxChars)}\n\n[... ${truncatedChars} more characters truncated]`;
};

export const serialize_conversation = (messages: message[]): string => {
  const parts: string[] = [];

  for (const msg of messages) {
    if (msg.role === "user") {
      const content = content_string(msg);
      if (content !== "") {
        parts.push(`[User]: ${content}`);
      }
    } else if (msg.role === "assistant") {
      const textParts: string[] = [];
      const thinkingParts: string[] = [];
      const toolCalls: string[] = [];

      if (msg.reasoning_content !== undefined && msg.reasoning_content !== "") {
        thinkingParts.push(msg.reasoning_content);
      } else if (msg.reasoning_details !== undefined && msg.reasoning_details !== "") {
        thinkingParts.push(msg.reasoning_details);
      } else if (msg.reasoning !== undefined && msg.reasoning !== "") {
        thinkingParts.push(msg.reasoning);
      }

      const content = content_string(msg);
      if (content !== "") {
        textParts.push(content);
      }

      for (const tc of msg.tool_calls ?? []) {
        let args: any = null;
        try {
          args = JSON.parse(tc.function.arguments);
        } catch {}

        let argsStr = "";
        if (args && typeof args === "object" && !Array.isArray(args)) {
          const kv: string[] = [];
          const keys = Object.keys(args).sort();
          for (const k of keys) {
            kv.push(`${k}=${safe_json_stringify(args[k])}`);
          }
          argsStr = kv.join(", ");
        } else {
          argsStr = tc.function.arguments;
        }
        toolCalls.push(`${tc.function.name}(${argsStr})`);
      }

      if (thinkingParts.length > 0) {
        parts.push(`[Assistant thinking]: ${thinkingParts.join("\n")}`);
      }
      if (textParts.length > 0) {
        parts.push(`[Assistant]: ${textParts.join("\n")}`);
      }
      if (toolCalls.length > 0) {
        parts.push(`[Assistant tool calls]: ${toolCalls.join("; ")}`);
      }
    } else if (msg.role === "tool" || msg.role === "toolResult") {
      const content = content_string(msg);
      if (content !== "") {
        parts.push(`[Tool result]: ${truncate_for_summary(content, 2000)}`);
      }
    }
  }

  return parts.join("\n\n");
};

export const bash_execution_to_text = (msg: any): string => {
  let text = `Ran \`${msg.command || ""}\`\n`;
  if (msg.output) {
    text += `\`\`\`\n${msg.output}\n\`\`\``;
  } else {
    text += "(no output)";
  }
  if (msg.cancelled) {
    text += "\n\n(command cancelled)";
  } else if (msg.exitCode !== null && msg.exitCode !== undefined && msg.exitCode !== 0) {
    text += `\n\nCommand exited with code ${msg.exitCode}`;
  }
  if (msg.truncated && msg.fullOutputPath) {
    text += `\n\n[Output truncated. Full output: ${msg.fullOutputPath}]`;
  }
  return text;
};

export const convert_to_llm = (messages: message[]): message[] => {
  const out: message[] = [];
  for (const m of messages) {
    switch (m.role) {
      case "bashExecution": {
        const anyMsg = m as any;
        if (anyMsg.excludeFromContext) {
          break;
        }
        out.push({
          role: "user",
          content: string_content(bash_execution_to_text(m)),
        });
        break;
      }
      case "custom": {
        out.push({
          role: "user",
          content: m.content,
        });
        break;
      }
      case "compactionSummary": {
        const content = CompactionSummaryPrefix + content_string(m) + CompactionSummarySuffix;
        out.push({
          role: "user",
          content: string_content(content),
        });
        break;
      }
      case "branchSummary": {
        const content = BranchSummaryPrefix + content_string(m) + BranchSummarySuffix;
        out.push({
          role: "user",
          content: string_content(content),
        });
        break;
      }
      default:
        out.push(m);
        break;
    }
  }
  return out;
};

export const calculate_context_tokens = (u: usage): number => {
  if (u.totalTokens !== undefined && u.totalTokens > 0) {
    return u.totalTokens;
  }
  return (u.input ?? 0) + (u.output ?? 0) + (u.cacheRead ?? 0) + (u.cacheWrite ?? 0);
};

export const get_last_assistant_usage = (entries: file_entry[]): usage | null => {
  for (let i = entries.length - 1; i >= 0; i--) {
    const entry = entries[i];
    if (entry.type === type_message && entry.message && entry.message.role === "assistant") {
      if (entry.message.usage) {
        return entry.message.usage;
      }
    }
  }
  return null;
};

export type context_usage_estimate = {
  tokens: number;
  usageTokens: number;
  trailingTokens: number;
  lastUsageIndex: number;
};

const ESTIMATED_IMAGE_CHARS = 4800;

export const estimate_text_and_image_content_chars = (msg: message): number => {
  const blocks = content_blocks(msg);
  if (!blocks || blocks.length === 0) {
    return content_string(msg).length;
  }
  let chars = 0;
  for (const block of blocks) {
    if (block.type === "text" && block.text) {
      chars += block.text.length;
    } else if (block.type === "image" || block.type === "image_url") {
      chars += ESTIMATED_IMAGE_CHARS;
    }
  }
  return chars;
};

export const estimate_message_tokens = (msg: message): number => {
  let chars = 0;

  switch (msg.role) {
    case "user": {
      chars = estimate_text_and_image_content_chars(msg);
      return Math.ceil(chars / 4);
    }
    case "assistant": {
      const text = content_string(msg);
      chars += text.length;
      if (msg.reasoning_content !== undefined) {
        chars += msg.reasoning_content.length;
      } else if (msg.reasoning_details !== undefined) {
        chars += msg.reasoning_details.length;
      } else if (msg.reasoning !== undefined) {
        chars += msg.reasoning.length;
      }
      for (const tc of msg.tool_calls ?? []) {
        chars += tc.function.name.length + safe_json_stringify(tc.function.arguments).length;
      }
      return Math.ceil(chars / 4);
    }
    case "custom":
    case "tool":
    case "toolResult": {
      chars = estimate_text_and_image_content_chars(msg);
      return Math.ceil(chars / 4);
    }
    case "bashExecution": {
      const anyMsg = msg as any;
      const cmd = typeof anyMsg.command === "string" ? anyMsg.command : "";
      const out = typeof anyMsg.output === "string" ? anyMsg.output : "";
      chars = cmd.length + out.length;
      return Math.ceil(chars / 4);
    }
    case "branchSummary":
    case "compactionSummary": {
      chars = content_string(msg).length;
      return Math.ceil(chars / 4);
    }
  }

  return 0;
};

export const estimate_tokens = (entry: file_entry): number => {
  if (entry.type === type_compaction || entry.type === type_branch_summary) {
    return Math.ceil((entry.summary ?? "").length / 4);
  }
  if (entry.type !== type_message || !entry.message) {
    return 0;
  }
  return estimate_message_tokens(entry.message);
};

export const estimate_context_tokens = (messages: message[]): context_usage_estimate => {
  let lastUsageIdx = -1;
  let lastUsage: usage | undefined;
  for (let i = messages.length - 1; i >= 0; i--) {
    if (messages[i].role === "assistant" && messages[i].usage) {
      lastUsageIdx = i;
      lastUsage = messages[i].usage;
      break;
    }
  }

  if (lastUsageIdx === -1) {
    let estimated = 0;
    for (const msg of messages) {
      estimated += estimate_message_tokens(msg);
    }
    return {
      tokens: estimated,
      usageTokens: 0,
      trailingTokens: estimated,
      lastUsageIndex: -1,
    };
  }

  const usageTokens = calculate_context_tokens(lastUsage!);
  let trailingTokens = 0;
  for (let i = lastUsageIdx + 1; i < messages.length; i++) {
    trailingTokens += estimate_message_tokens(messages[i]);
  }

  return {
    tokens: usageTokens + trailingTokens,
    usageTokens,
    trailingTokens,
    lastUsageIndex: lastUsageIdx,
  };
};

export const should_compact = (
  contextTokens: number,
  contextWindow: number,
  settings: compaction_settings,
): boolean => {
  if (!settings.enabled) {
    return false;
  }
  return contextTokens > contextWindow - settings.reserve_tokens;
};

const find_valid_cut_points = (
  entries: file_entry[],
  startIndex: number,
  endIndex: number,
): number[] => {
  const cutPoints: number[] = [];
  for (let i = startIndex; i < endIndex; i++) {
    const entry = entries[i];
    if (entry.type === type_message && entry.message) {
      const role = entry.message.role;
      if (
        role === "user" ||
        role === "assistant" ||
        role === "bashExecution" ||
        role === "custom" ||
        role === "branchSummary" ||
        role === "compactionSummary"
      ) {
        cutPoints.push(i);
      }
    } else if (entry.type === type_branch_summary || entry.type === type_custom_message) {
      cutPoints.push(i);
    }
  }
  return cutPoints;
};

const find_turn_start_index = (
  entries: file_entry[],
  entryIndex: number,
  startIndex: number,
): number => {
  for (let i = entryIndex; i >= startIndex; i--) {
    const entry = entries[i];
    if (entry.type === type_branch_summary || entry.type === type_custom_message) {
      return i;
    }
    if (entry.type === type_message && entry.message) {
      const role = entry.message.role;
      if (role === "user" || role === "bashExecution") {
        return i;
      }
    }
  }
  return -1;
};

export type cut_point_result = {
  firstKeptEntryIndex: number;
  turnStartIndex: number;
  isSplitTurn: boolean;
};

export const find_cut_point = (
  entries: file_entry[],
  startIndex: number,
  endIndex: number,
  keepRecentTokens: number,
): cut_point_result => {
  const cutPoints = find_valid_cut_points(entries, startIndex, endIndex);
  if (cutPoints.length === 0) {
    return { firstKeptEntryIndex: startIndex, turnStartIndex: -1, isSplitTurn: false };
  }

  let accumulatedTokens = 0;
  let cutIndex = cutPoints[0];

  for (let i = endIndex - 1; i >= startIndex; i--) {
    const entry = entries[i];
    if (entry.type !== type_message) {
      continue;
    }
    accumulatedTokens += estimate_tokens(entry);
    if (accumulatedTokens >= keepRecentTokens) {
      for (const cp of cutPoints) {
        if (cp >= i) {
          cutIndex = cp;
          break;
        }
      }
      break;
    }
  }

  while (cutIndex > startIndex) {
    const prevEntry = entries[cutIndex - 1];
    if (prevEntry.type === type_compaction || prevEntry.type === type_message) {
      break;
    }
    cutIndex--;
  }

  const cutEntry = entries[cutIndex];
  const isUserMessage =
    cutEntry.type === type_message &&
    cutEntry.message &&
    cutEntry.message.role === "user";
  let turnStartIndex = -1;
  if (!isUserMessage) {
    turnStartIndex = find_turn_start_index(entries, cutIndex, startIndex);
  }

  return {
    firstKeptEntryIndex: cutIndex,
    turnStartIndex,
    isSplitTurn: !isUserMessage && turnStartIndex !== -1,
  };
};

export const get_latest_compaction_entry = (branch: file_entry[]): file_entry | null => {
  for (let i = branch.length - 1; i >= 0; i--) {
    if (branch[i].type === type_compaction) {
      return branch[i];
    }
  }
  return null;
};

/*
PORT STATUS
source path: backend/session/compaction_utils.go
source lines: 410
confidence: high
status: phase_b_compile
*/
